pennylane/measurements/classical_shadow.py

```diff
@@ -314,6 +315,106 @@ def process(self, tape, device):
+        num_dev_qubits = len(state.shape)
+
+        # seed the random measurement generation so that recipes
+        # are the same for different executions with the same seed
+        recipe_rng = np.random.RandomState(self.seed)
+        recipes = recipe_rng.randint(0, 3, size=(shots, n_qubits))
+
+        bit_rng = np.random.default_rng(rng)
+
+        obs_list = np.stack(
+            [
+                qml.PauliX.compute_matrix(),
+                qml.PauliY.compute_matrix(),
+                qml.PauliZ.compute_matrix(),
+            ]
+        )
+
+        # the diagonalizing matrices corresponding to the Pauli observables above
+        diag_list = np.stack(
+            [
+                qml.Hadamard.compute_matrix(),
+                qml.Hadamard.compute_matrix() @ qml.RZ.compute_matrix(-np.pi / 2),
+                qml.Identity.compute_matrix(),
+            ]
+        )
+        obs = obs_list[recipes]
+        uni = diag_list[recipes]
+
+        # There's a significant speedup if we use the following iterative
+        # process to perform the randomized Pauli measurements:
+        #   1. Randomly generate Pauli observables for all snapshots for
+        #      a single qubit (e.g. the first qubit).
+        #   2. Compute the expectation of each Pauli observable on the first
+        #      qubit by tracing out all other qubits.
+        #   3. Sample the first qubit based on each Pauli expectation.
+        #   4. For all snapshots, determine the collapsed state of the remaining
+        #      qubits based on the sample result.
+        #   4. Repeat iteratively until no qubits are remaining.
+        #
+        # Observe that after the first iteration, the second qubit will become the
+        # "first" qubit in the process. The advantage to this approach as opposed to
+        # simulataneously computing the Pauli expectations for each qubit is that
+        # the partial traces are computed over iteratively smaller subsystems, leading
+        # to a significant speed-up.
+
+        # transpose the state so that the measured wires appear first
+        unmeasured_wires = [i for i in range(num_dev_qubits) if i not in mapped_wires]
+        transposed_state = np.transpose(state, axes=mapped_wires.tolist() + unmeasured_wires)
+
+        outcomes = np.zeros((shots, n_qubits))
+        stacked_state = np.stack([transposed_state for _ in range(shots)])
+
+        for active_qubit in range(n_qubits):
+            # trace out every qubit except the first
+            first_qubit_state = np.einsum(
+                f"{ABC[num_dev_qubits - active_qubit + 1]}{ABC[:num_dev_qubits - active_qubit]},"
+                f"{ABC[num_dev_qubits - active_qubit + 1]}{ABC[num_dev_qubits - active_qubit]}{ABC[1:num_dev_qubits - active_qubit]}"
+                f"->{ABC[num_dev_qubits - active_qubit + 1]}a{ABC[num_dev_qubits - active_qubit]}",
+                stacked_state,
+                np.conj(stacked_state),
+            )
```

Comment:
```suggestion
            # stacked_state loses a dimension each loop

            # trace out every qubit except the first
            num_remaining_qubits = num_dev_qubits - active_qubit
            conj_state_first_qubit = ABC[num_remaining_qubits]
            stacked_dim = ABC[num_remaining_qubits + 1]

            state_str = f"{stacked_dim}{ABC[:num_remaining_qubits]}"
            conj_state_str = f"{stacked_dim}{conj_state_first_qubit}{ABC[1:num_remaining_qubits]}"
            target_str = f"{stacked_dim}a{conj_state_first_qubit}"

            first_qubit_state = np.einsum(
                f"{state_str},{conj_state_str}->{target_str}",
                stacked_state,
                np.conj(stacked_state),
            )
```

This might clear up what's happening in this block a little bit.